snow/engine/snowman/event/queue.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package event
+
+import (
+	"context"
+
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+type Job interface {
+	Execute(context.Context) error
+	Cancel(context.Context) error
+}
+
+type job[T comparable] struct {
+	// If empty, the job is ready to be executed.
+	dependencies set.Set[T]
+	// If true, the job should be cancelled.
+	shouldCancel bool
+	// If nil, the job has already been executed or cancelled.
+	job Job
+}
+
+type Queue[T comparable] struct {
```

Comment:
(pedantry) It's not really a queue, which might give the wrong impression to someone who wants to use it in the future. The same could probably be argued about the package name too: `event.Queue` is misleading. That said, I'm not exactly sure what it is! Perhaps rename both and this becomes a `job.Pool`?

Reply:
Hm - perhaps a job.Scheduler?

Reply:
I like it